### PR TITLE
Props compile

### PR DESCRIPTION
**Pull request by harlouci**
All __props__ checked in folder theano/compile/

### CODE REVIEW

File: theano/compile/builders.py [hunk 1]
```diff
@@ -101,14 +102,6 @@ def __init__(self, inputs, outputs, **kwargs):
         self.input_types = [input.type for input in inputs]
         self.output_types = [output.type for output in outputs]
 
-    def __eq__(self, other):
-        # TODO: recognize a copy
-        return self is other
```
Comment (nouiz) [inline theano/compile/builders.py]: Keep \__eq__, it don't do the same as the default. And keep the comment. Also keep \__hash__ and its todo. remove \__props__ for this op, it is a special op.

File: theano/compile/ops.py [hunk 2]
```diff
@@ -623,6 +603,7 @@ class Rebroadcast(gof.Op):
     c_code_and_version = {}
 
     check_input = False
+    __props__ = ("axis")
 
     def __init__(self, *axis):
         self.axis = dict(axis)
```
Comment (nouiz) [inline theano/compile/ops.py]: In \__hash__ axis was sorted. I think we should use an OrderedDict, that we create empty and add the axis to it in the sorted order. That would make this Op better and allow to use \__props__ without regression.

File: theano/compile/ops.py [hunk 3]
```diff
@@ -488,13 +475,6 @@ def __init__(self, fn, itypes, otypes, infer_shape):
         if self.__infer_shape is not None:
             self.infer_shape = self._infer_shape
 
-    def __eq__(self, other):
-        return (type(self) == type(other) and
-                self.__fn == other.__fn)
-
-    def __hash__(self):
-        return hash(type(self)) ^ hash(self.__fn)
-
```
Comment (nouiz) [inline theano/compile/ops.py]: This is a special op. itypes and otypes are only used when building the graph, so they aren't "props" like other props. For now, don't change this op. We are missing what is needed to make it work.

File: theano/compile/ops.py (hunk 2 above)
Comment (harlouci) [inline theano/compile/ops.py]: Could you be more explicit, please. I don't figure out the solution.

Comment (nouiz) [inline theano/compile/ops.py]: In the init, do this:
```
# Sort them to make sure we merge all possible case.
items = sorted(iteritems(self.axis))
self.axis = OrderedDict(items)
```

and use this import `from theano.compat import OrderedDict`

**Comment by nouiz** [comment]
travis fail.

**Comment by harlouci** [comment]
Inserting

 # Sort them to make sure we merge all possible case.
        items = sorted(iteritems(self.axis))
        self.axis = OrderedDict(items)

was your idea.

What is wrong about that ?

2015-07-17 9:21 GMT-04:00 Frédéric Bastien <notifications@github.com>:

**Comment by lamblin** [comment]
@harlouci `self.axis` has not been assigned yet. You would have to use `axis` instead on the first line.

File: theano/compile/ops.py [hunk 4]
```diff
@@ -623,21 +604,16 @@ class Rebroadcast(gof.Op):
     c_code_and_version = {}
 
     check_input = False
+    __props__ = ("axis",)
 
     def __init__(self, *axis):
-        self.axis = dict(axis)
+        # Sort them to make sure we merge all possible case.
+        items = sorted(iteritems(axis))
```
Comment (lamblin) [inline theano/compile/ops.py]: Actually, axis is a list of pairs, so you cannot call iteritems on it.
I think that should be:
```python
items = sorted(axis)
self.axis = OrderedDict(items)
```
but please try it first.

Comment (harlouci) [inline theano/compile/ops.py]: theano-nose theano/compile/tests/test_ops.py

says OK

**Comment by lamblin** [comment]
So OrderedDict is not hashable either, so it cannot be part of props.
One solution would be to have two attributes: one is a list of pairs, one an OrderedDict (or even a dict, like currently). Only the list of pairs would be part of `__props__`, but the OrderedDict would be used at the other places.

**Comment by nouiz** [comment]
I used a different strategy, now I enabled the hash of dict in props. What do you think of this? I think it is more general and easier to do.

**Comment by nouiz** [comment]
I updated the PR.

**Pull request merged**